src/install/install.ts

```diff
@@ -0,0 +1,148 @@
+ * output directory.
+ */
+async function _install(offline: boolean, bowerJson?: JsonObject, componentDirectory?: string, variantName?: string): Promise<void> {
+  const config = defaultBowerConfig({
+    save: false,
+    directory: componentDirectory,
+    offline,
+  });
+
+  const bowerLogger = new BowerLogger();
+  const cwd: string = config.cwd || process.cwd();
+  const renderer = new StandardRenderer('install', {
+    cwd,
+    color: true,
+  });
+  bowerLogger.on('log', (log: bower.LogData) => renderer.log(log));
+  bowerLogger.on('end', (data: any) => renderer.end(data));
+  bowerLogger.on('error', (err: Error) => renderer.error(err));
+
+  const project = new BowerProject(config, bowerLogger);
+
+  // This is the only way I could find to provide a JSON object to the
+  // Project. It's a hack, and might break in the future, but it works.
+  if (bowerJson) {
+    (project as any)._json = bowerJson;
+    // Generate a new fake bwoer.json name because Bower is insting on
+    // overwriting this file, even with the {save: false}.
+    // TODO(justinfagnani): Figure this out
+    const fileName = variantName ? `bower-${variantName}.json` : `bower.json`;
+    (project as any)._jsonFile = path.join(cwd, fileName);
+  }
+
+  await project.install([], {save: false, offline}, config);
+}
+
+async function installDefault(offline: boolean): Promise<void> {
+  logger.info(`Installing default Bower components...`);
+  await _install(offline);
+  logger.info(`Finished installing default Bower components`);
+}
+
+async function installVariants(offline: boolean): Promise<void> {
+  const bowerJson = await new Promise<any>((resolve, reject) => {
+    const config = defaultBowerConfig({
+      save: false,
+    });
+    const cwd = config.cwd || process.cwd();
+    readBowerJson(cwd, {}, (err: any, json: any) => {
+      err ? reject(err) : resolve(json);
+    });
+  });
+
+  // Variants are patches ontop of the default bower.json, typically used to
+  // override dependencies to specific versions for testing. Variants are
+  // installed into folders named "bower_components-{variantName}", which are
+  // used by other tools like polyserve.
+  const variants = bowerJson['variants'];
+  if (variants) {
+    await Promise.all(Object.keys(variants).map(async (variantName) => {
+      const variant = variants[variantName];
+      const variantBowerJson = _mergeJson(variant, bowerJson) as JsonObject;
+      const variantDirectory = `bower_components-${variantName}`;
+      logger.info(`Installing variant ${variantName} to ${variantDirectory}...`);
+      await _install(offline, variantBowerJson, variantDirectory, variantName);
+      logger.info(`Finished installing variant ${variantName}`);
+    }));
+  }
+}
+
+/**
+ * Exported only for testing
+ */
+export function _mergeJson(from: JsonValue, to: JsonValue): JsonValue {
+  if (isPrimitiveOrArray(from) || isPrimitiveOrArray(to)) {
```

Comment:
Continuing the conversation here since github is crazy and won't let me comment in original thread.

> What would the implementation of isJsonObject look like?

`return typeof x === 'object' && !Array.isArray(x)` (You can add a check for null in there as well, I wasn't sure if that was intentionally meant to be run through `mergeJson()` in your implementation or not.)